### PR TITLE
Implement new L2 attack

### PR DESCRIPTION
**Pull request by carlini**
New L2 attack implementation from "Towards Evaluating the Robustness of Neural Networks".

Commits are cherry-picked from #156.

### CODE REVIEW

File: cleverhans/attacks.py [hunk 1]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        self.structural_kwargs = ['nb_classes',
+                                  'batch_size', 'confidence',
+                                  'targeted', 'learning_rate',
+                                  'binary_search_steps', 'max_iterations',
+                                  'abort_early', 'initial_const',
+                                  'clip_min', 'clip_max']
+
+    def generate(self, x, **kwargs):
+        """
+        Generate adversarial samples and return them in a Numpy array.
```
Comment (npapernot) [inline cleverhans/attacks.py]: The `generate` method should return a tensor, this comment is likely to be misleading for users (even if this implementation is wrapped).

File: cleverhans/attacks.py [hunk 2]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+
+        :param x: (required) A Numpy array with the original inputs.
```
Comment (npapernot) [inline cleverhans/attacks.py]: See above

File: cleverhans/attacks.py [hunk 3]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        :param y: (optional) A Numpy array with the labels that we either
```
Comment (npapernot) [inline cleverhans/attacks.py]: See above

File: cleverhans/attacks.py [hunk 4]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+                  should target (if targeted=True) or avoid (if
+                  target=False). If None, use labels the classifier assigns.
+        :param nb_classes: The number of classes the model has.
```
Comment (npapernot) [inline cleverhans/attacks.py]: Could this be inferred from the model object? (to remove the number of arguments that users need to pass.

File: cleverhans/attacks.py [hunk 5]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        :param confidence: Confidence of adversarial examples: higher produces
+                           examples that are farther away, but more strongly
```
Comment (npapernot) [inline cleverhans/attacks.py]: "farther away" is somewhat vague. Could we instead say something along the lines of "perturbations with a higher L2 norm"

File: cleverhans/attacks.py [hunk 6]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
         return True
 
 
+class CarliniWagnerL2(Attack):
```
Comment (npapernot) [inline cleverhans/attacks.py]: Given that you will make PRs for the L0 and Linf attacks, is it possible to make the class general enough, that it can take a norm argument and call the right underlying algorithm?  Or is that not possible because the attacks have very different interfaces for different norms?

File: cleverhans/attacks.py [hunk 7]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        :param x: (required) A Numpy array with the original inputs.
+        :param y: (optional) A Numpy array with the labels that we either
+                  should target (if targeted=True) or avoid (if
+                  target=False). If None, use labels the classifier assigns.
+        :param nb_classes: The number of classes the model has.
+        :param confidence: Confidence of adversarial examples: higher produces
+                           examples that are farther away, but more strongly
+                           classified as adversarial.
+        :param batch_size: Number of attacks to run simultaneously.
+        :param targeted: True if we should perform a targetted attack, False
```
Comment (npapernot) [inline cleverhans/attacks.py]: Is the spelling for targetted, targeted or targetted? (I'm actually not sure :) )

File: cleverhans/attacks.py [hunk 8]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+                         otherwise.
+        :param learning_rate: The learning rate for the attack algorithm.
+                              Smaller values produce better results but are
+                              slower to converge.
+        :param binary_search_steps: The number of times we perform binary
+                                    search to find the optimal tradeoff-
+                                    constant between distance and confidence.
```
Comment (npapernot) [inline cleverhans/attacks.py]: Would "norm of the perturbation" and "confidence of the misclassification" make that comment easier to understand for newcomers?

File: cleverhans/attacks.py [hunk 9]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        :param max_iterations: The maximum number of iterations. Larger values
+                               are more accurate; setting too small will
+                               require a large learning rate and will produce
```
Comment (npapernot) [inline cleverhans/attacks.py]: It sounds like something is missing from the comparison. Do you mean that to compensate with the smaller number of iterations, the optimization has to take large steps at each iteration (and hence a larger learning rate) to be as likely to succeed (even though in practice it doesn't)?

File: cleverhans/attacks.py [hunk 10]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+                  should target (if targeted=True) or avoid (if
+                  target=False). If None, use labels the classifier assigns.
+        :param nb_classes: The number of classes the model has.
+        :param confidence: Confidence of adversarial examples: higher produces
+                           examples that are farther away, but more strongly
+                           classified as adversarial.
+        :param batch_size: Number of attacks to run simultaneously.
+        :param targeted: True if we should perform a targetted attack, False
+                         otherwise.
+        :param learning_rate: The learning rate for the attack algorithm.
```
Comment (npapernot) [inline cleverhans/attacks.py]: Is learning rate the best name to call this parameter? I am wondering if this will be confusing for new users (because of the learning rate used during training)? 

I don't have a strong opinion however.

File: cleverhans/attacks.py [hunk 11]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+                              slower to converge.
+        :param binary_search_steps: The number of times we perform binary
+                                    search to find the optimal tradeoff-
+                                    constant between distance and confidence.
+        :param max_iterations: The maximum number of iterations. Larger values
+                               are more accurate; setting too small will
+                               require a large learning rate and will produce
+                               poor results.
+        :param abort_early: If true, allows early aborts if gradient descent
+                            gets stuck.
```
Comment (npapernot) [inline cleverhans/attacks.py]: Can you define a little more precisely what you mean by "get stuck". I suppose it means that the gradients are vanishing for several iterations?

File: cleverhans/attacks.py [hunk 12]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        :param initial_const: The initial tradeoff-constant to use to tune the
+                              relative importance of distance and confidence.
```
Comment (npapernot) [inline cleverhans/attacks.py]: See above (regarding "distance" and "confidence")

File: cleverhans/attacks.py [hunk 13]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+                      x.get_shape().as_list()[1:])
+
+        if 'y' in kwargs:
+            labels = kwargs['y']
+        else:
+            if self.targeted:
+                raise ValueError("Must supply target labels 'y'"
+                                 " in targeted attack.")
+            # TODO abstract this out for other classes too
+            preds = self.model(x)
```
Comment (npapernot) [inline cleverhans/attacks.py]: Should this be `model.get_probs()` with the new interface?

File: cleverhans/attacks.py [hunk 14]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        return wrap
+
+    def parse_params(self, y=None, y_val=None, nb_classes=10,
+                     batch_size=1, confidence=0,
+                     targeted=True, learning_rate=5e-3,
+                     binary_search_steps=5, max_iterations=1000,
+                     abort_early=True, initial_const=1e-2,
+                     clip_min=0, clip_max=1):
+
+        # ignore the y and y_val argument
```
Comment (npapernot) [inline cleverhans/attacks.py]: Could you elaborate on why these two arguments are ignored? Not necessarily in the code, but in this thread at least.

File: cleverhans/attacks.py (hunk 2 above)
Comment (goodfeli) [inline cleverhans/attacks.py]: Not important enough to delay merge, but preferred capitalization seems to be NumPy

File: cleverhans/attacks_tf.py [hunk 15]
```diff
@@ -417,3 +417,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
 
     # Return augmented training data (needs to be labeled afterwards)
     return X_sub
+
+
+class CarliniWagnerL2:
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: I think we already talked about that, but I'm wondering how we should organize the code here. This is the first attack that has an underlying class object. My question is do we want to eventually merge this `CarliniWagnerL2` class in `attacks_tf.py` w/ the wrapper class `CarliniWagnerL2` in `attacks.py`.

The reason this might make sense is that once we drop support for Theano in a few months, the wrapper classes in `attacks.py` will be pretty bare bones, so putting the code all in one place might make things easier to maintain (we could also apply the same logic to the other attacks).

Not sure what is the best way to go, @goodfeli do you have any thoughts?

File: cleverhans/attacks_tf.py [hunk 16]
```diff
@@ -417,3 +417,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
...
+    def __init__(self, sess, model, batch_size, confidence,
+                 targeted, learning_rate,
+                 binary_search_steps, max_iterations,
+                 abort_early, initial_const,
+                 clip_min, clip_max, num_labels, shape):
+        """
+        This attack was originally proposed by Carlini and Wagner. It is an
+        iterative attack that finds adversarial examples on many defenses that
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: See above

File: cleverhans/attacks.py (hunk 3 above)
Comment (goodfeli) [inline cleverhans/attacks.py]: It seems kind of dangerous to have this option take on almost opposite meanings based on the values of other arguments.
Would it make sense to have target be either None or a tensor containing the target? Then raise an error if both y and target are specified.

File: cleverhans/attacks.py [hunk 17]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
         return True
 
 
+class CarliniWagnerL2(Attack):
+    """
+    This attack was originally proposed by Carlini and Wagner. It is an
+    iterative attack that finds adversarial examples on many defenses that
```
Comment (npapernot) [inline cleverhans/attacks.py]: Could you add a high-level TL;DR of what the attack is doing? Something like running Adam on customized loss functions

File: cleverhans/attacks.py [hunk 18]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+                                  'clip_min', 'clip_max']
+
+    def generate(self, x, **kwargs):
+        """
+        Generate adversarial samples and return them in a Numpy array.
+
+        :param x: (required) A Numpy array with the original inputs.
+        :param y: (optional) A Numpy array with the labels that we either
+                  should target (if targeted=True) or avoid (if
+                  target=False). If None, use labels the classifier assigns.
```
Comment (goodfeli) [inline cleverhans/attacks.py]: should be "targeted" not "target" if keeping the current names

File: cleverhans/attacks_tf.py [hunk 19]
```diff
@@ -417,3 +417,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
...
+        are robust to other attacks.
+        Paper link: https://arxiv.org/abs/1608.04644
+
+        :param confidence: Confidence of adversarial examples: higher produces
+                           examples that are farther away, but more strongly
+                           classified as adversarial.
+        :param batch_size: Number of attacks to run simultaneously.
+        :param targeted: True if we should perform a targetted attack, False
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Typo? See above

File: cleverhans/attacks_tf.py [hunk 20]
```diff
@@ -417,3 +417,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
...
+                         otherwise.
+        :param learning_rate: The learning rate for the attack algorithm.
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: See above, and this probably applies to most of the docstring since it is shared w/ the wrapper class in `attacks.py`

File: cleverhans/attacks.py [hunk 21]
```diff
@@ -514,6 +514,115 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        :param nb_classes: The number of classes the model has.
+        :param confidence: Confidence of adversarial examples: higher produces
```
Comment (goodfeli) [inline cleverhans/attacks.py]: Is this a minimum confidence level after which the optimization stops? Or a target confidence level the optimization aims at?

File: cleverhans/attacks.py (hunk 7 above)
Comment (goodfeli) [inline cleverhans/attacks.py]: typo "targetted"

File: tests_tf/test_mnist_tutorial_jsma.py [hunk 22]
```diff
@@ -16,7 +16,7 @@ def test_mnist_tutorial_jsma(self):
         jsma_tutorial_args = {'train_start': 0,
                               'train_end': 1000,
                               'test_start': 0,
-                              'test_end': 1666,
+                              'test_end': 166,
```
Comment (npapernot) [inline tests_tf/test_mnist_tutorial_jsma.py]: Why was this changed?

**Comment by npapernot** [comment]
Thanks a lot for the PR Nicholas, I left a few comments for the first part of the code and will take a look again when you got a chance to make the edits (but I suspect the other files are more straightforward from the quick glance I took).

File: cleverhans/attacks_tf.py (hunk 15 above)
Comment (goodfeli) [inline cleverhans/attacks_tf.py]: What's the purpose of having two different classes?

Also, if we keep the two-class version, I suspect this one should inherit from `object` to make it a "new-style class".

File: tests_tf/test_attacks.py [hunk 23]
```diff
@@ -112,7 +113,7 @@ def test_generate_np(self):
         perturbation_norm = np.sqrt(np.sum(perturbation**2, axis=1))
         # test perturbation norm
         self.assertTrue(np.allclose(perturbation_norm, self.attack.eps))
-
+
```
Comment (goodfeli) [inline tests_tf/test_attacks.py]: probably better to not add this but not a big deal

File: tutorials/mnist_tutorial_cw.py [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+    # Instantiate a CW attack object
+    cw = CarliniWagnerL2(model, back='tf', sess=sess)
+
+    idxs = [np.where(np.argmax(Y_test, axis=1) == i)[0][0] for i in range(10)]
+    if targeted:
+        # Initialize our array for grid visualization
+        grid_shape = (nb_classes, nb_classes, img_rows, img_cols, channels)
+        grid_viz_data = np.zeros(grid_shape, dtype='f')
+
+        onehot = np.zeros((10, 10))
```
Comment (goodfeli) [inline tutorials/mnist_tutorial_cw.py]: maybe `one_hot` just for consistent pep8 naming scheme / matching tf.one_hot

File: cleverhans/attacks.py (hunk 3 above)
Comment (carlini) [inline cleverhans/attacks.py]: Okay, makes sense. Should we standardize on the name of the y values we supply? JSMA calls it "targets", FGM calls it "y". Should I use "targets" for the targeted mode and "y" for untargeted?

File: cleverhans/attacks.py (hunk 6 above)
Comment (carlini) [inline cleverhans/attacks.py]: Yeah, there are different interfaces (e.g., there's no binary search in l0/li), but those have other constants l2 doesn't.

File: cleverhans/attacks.py (hunk 1 above)
Comment (carlini) [inline cleverhans/attacks.py]: Ah sorry, yeah, I copied the docstring and didn't update this.

File: cleverhans/attacks.py (hunk 7 above)
Comment (carlini) [inline cleverhans/attacks.py]: Apparently neither was I. But targeted is right.

File: cleverhans/attacks.py (hunk 10 above)
Comment (carlini) [inline cleverhans/attacks.py]: I don't have a better word, do you? I use learning_rate because that's what it's used for. step_size isn't entirely accurate since adam does adaptive step sizes for a given learning rate.

File: cleverhans/attacks.py (hunk 9 above)
Comment (carlini) [inline cleverhans/attacks.py]: What I mean to say: If the number of iterations is small, then the learning rate must be high. This in turn makes the distance to adversarial examples higher than it must be.

I'll try to edit the docs.

File: cleverhans/attacks.py (hunk 11 above)
Comment (carlini) [inline cleverhans/attacks.py]: Will do. That's what I mean.

File: cleverhans/attacks.py (hunk 14 above)
Comment (carlini) [inline cleverhans/attacks.py]: Because I specifically handle them in the generate() method, and don't do anything with it in the parse_params. I might be able to restructure the code to do it in parse_params() now that generate_np() is implemented in the Attack class.

File: tests_tf/test_mnist_tutorial_jsma.py (hunk 22 above)
Comment (carlini) [inline tests_tf/test_mnist_tutorial_jsma.py]: By accident -- cherrypicked it it seems.

File: cleverhans/attacks_tf.py (hunk 15 above)
Comment (carlini) [inline cleverhans/attacks_tf.py]: Yeah, I should remove the second class. It was a class so that I could easily deal with graph caching, but now that generate_np is rewritten, I should clean this up.

File: cleverhans/attacks.py (hunk 21 above)
Comment (carlini) [inline cleverhans/attacks.py]: It targets specific the confidence that you set. So if you set 0, it makes it just cross the decision boundary. If you set 5 then it wants exactly 5 (and reports "failure" if it's less than 5).

File: cleverhans/attacks_tf.py (hunk 15 above)
Comment (carlini) [inline cleverhans/attacks_tf.py]: I will kill this class. It's only a class because it was convenient when I was doing my own caching in the attack, but with the generate_np rewrite it should just be its own method.

File: cleverhans/attacks.py (hunk 17 above)
Comment (carlini) [inline cleverhans/attacks.py]: Done in f8feb2e

File: cleverhans/attacks.py (hunk 6 above)
Comment (npapernot) [inline cleverhans/attacks.py]: Ok, let's keep it that way in that case

File: cleverhans/attacks.py (hunk 10 above)
Comment (npapernot) [inline cleverhans/attacks.py]: I don't have a better suggestion, let's just keep it that way then. The rest of the sentence should make it clear enough.

File: cleverhans/attacks.py (hunk 9 above)
Comment (npapernot) [inline cleverhans/attacks.py]: Ok thanks!

File: cleverhans/attacks_tf.py (hunk 15 above)
Comment (npapernot) [inline cleverhans/attacks_tf.py]: That sounds good!

File: cleverhans/attacks.py [hunk 25]
```diff
@@ -540,35 +544,41 @@ def __init__(self, model, back='tf', sess=None):
...
+                               results. Using only a few iterations requires
+                               a larger learning rate, and will produce larger
+                               distortion results.
         :param abort_early: If true, allows early aborts if gradient descent
-                            gets stuck.
+                            is unable to make progress (i.e., gets stuck in
+                            a local minimum).
         :param initial_const: The initial tradeoff-constant to use to tune the
-                              relative importance of distance and confidence.
+                              relative importance of size of the pururbation
```
Comment (npapernot) [inline cleverhans/attacks.py]: There is a typo

**Comment by npapernot** [comment]
The tests are failing because of the two following `pep8` errors:

```
/home/travis/build/tensorflow/cleverhans/cleverhans/attacks.py:556:72: W291 trailing whitespace
/home/travis/build/tensorflow/cleverhans/cleverhans/attacks.py:577:77: W291 trailing whitespace
```

**Comment by carlini** [comment]
I have a fix for those pep8 errors, along with removing the attacks_tf.py extra class to make the code cleaner. If not done tonight, will be done tomorrow.

**Comment by goodfeli** [comment]
Re standardizing on "targeted", "y", etc., do we maybe want to make targeted and non-targeted attacks official part of the Attack interface?

File: cleverhans/attacks_tf.py (hunk 15 above)
Comment (carlini) [inline cleverhans/attacks_tf.py]: Okay, so I think that actually having the second class is useful. Since the attack is implemented in python, and generate() is implemented with a py_func, if the py_func created a second instance of the class, then the pseudo-code

    args = {...}
    xs = CarliniWagnerL2().generate(...)
    a = sess.run(xs, args)
    b = sess.run(xs, args)

would result in constructing the graph twice. JSMA has this problem too, but gets around it by placing a call to

    grads = jacobian_graph(preds, x, self.nb_classes)  

outside of the py_func. This is equivelant to what I do now, which is instantiate the second class outside of the py_func. In my case, it's easier to have a second class instead of a helper function because there are so many elements of the graph that are re-used and just saved as instance variables.

Comment (carlini) [inline cleverhans/attacks_tf.py]: See discussion under Ian's comment.

**Comment by carlini** [comment]
Rebase seems to have slightly confused googlebot.

Unless there's anything else, I'm happy with the current state of the commit. The final possible change I have in mind would be to remove nb_classes and just figure it out based on the model, but I think I will do that in a separate PR so I can fix JSMA too.

**Comment by npapernot** [comment]
Yes, let's keep the nb_classes removal for another PR. 

I'll take another pass at this one

**Comment by npapernot** [comment]
And yes the CLA bot bugs when someone rebases through the online interface, but we can still merge so it's OK

File: cleverhans/attacks.py [hunk 26]
```diff
@@ -571,6 +570,130 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+                                  'binary_search_steps', 'max_iterations',
+                                  'abort_early', 'initial_const',
+                                  'clip_min', 'clip_max']
+
+        if not isinstance(self.model, Model):
+            self.model = CallableModelWrapper(self.model, 'logits')
+        
+    def generate(self, x, **kwargs):
+        """
+        Return a tensor that consturcts adversarial examples for the given
```
Comment (npapernot) [inline cleverhans/attacks.py]: There is a typo

File: cleverhans/attacks.py [hunk 27]
```diff
@@ -571,6 +570,130 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        input. Generate uses tf.py_func in order to operate over tensors.
+
+        :param x: (required) A tensor with the original inputs.
+        :param y: (optional) A tensor with the true labels for an untargeted
+                  attack. If None (and y_target is None) then use the
+                  original labels the classifier assigns.
```
Comment (npapernot) [inline cleverhans/attacks.py]: Could we remove "original", it sounds like we are talking about the labels from the dataset?

File: cleverhans/attacks.py [hunk 28]
```diff
@@ -571,6 +570,130 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        :param y_target: (optional) A tensor with the true labels for an
```
Comment (npapernot) [inline cleverhans/attacks.py]: This seems to have been copied over

File: cleverhans/attacks.py [hunk 29]
```diff
@@ -571,6 +570,130 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+
+        attack = CWL2(self.sess, self.model, self.batch_size,
+                      self.confidence, 'y_target' in kwargs,
+                      self.learning_rate, self.binary_search_steps,
+                      self.max_iterations, self.abort_early,
+                      self.initial_const, self.clip_min, self.clip_max,
+                      self.nb_classes, x.get_shape().as_list()[1:])
+
+        if 'y' in kwargs and 'y_target' in kwargs:
+            raise ValueError("Can not send both 'y' and 'y_target'.")
```
Comment (npapernot) [inline cleverhans/attacks.py]: send -> set

File: cleverhans/attacks.py [hunk 30]
```diff
@@ -571,6 +570,130 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        elif 'y' in kwargs:
+            labels = kwargs['y']
+        elif 'y_target' in kwargs:
+            labels = kwargs['y_target']
+        else:
+            # TODO abstract this out for other classes too
```
Comment (npapernot) [inline cleverhans/attacks.py]: Could you raise an issue instead and remove the comment? It's easier to track what's open in the issue tracker

File: cleverhans/attacks.py [hunk 31]
```diff
@@ -571,6 +570,130 @@ def parse_params(self, eps=2.0, num_iterations=1, xi=1e-6, clip_min=None,
...
+        return wrap
+
+    def parse_params(self, y=None, y_target=None, nb_classes=10,
+                     batch_size=1, confidence=0,
+                     learning_rate=5e-3,
+                     binary_search_steps=5, max_iterations=1000,
+                     abort_early=True, initial_const=1e-2,
+                     clip_min=0, clip_max=1):
+
+        # ignore the y and y_val argument
```
Comment (npapernot) [inline cleverhans/attacks.py]: Anything new related to this?

File: cleverhans/attacks_tf.py [hunk 32]
```diff
@@ -437,3 +437,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
 
     # Return augmented training data (needs to be labeled afterwards)
     return X_sub
+
+
+class CarliniWagnerL2:
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Anything new related to @goodfeli 's comment on inheriting from `object`?

File: cleverhans/attacks_tf.py [hunk 33]
```diff
@@ -437,3 +437,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
...
+        This attack was originally proposed by Carlini and Wagner. It is an
+        iterative attack that finds adversarial examples on many defenses that
+        are robust to other attacks.
+        Paper link: https://arxiv.org/abs/1608.04644
+
+        :param confidence: Confidence of adversarial examples: higher produces
+                           examples that are farther away, but more strongly
+                           classified as adversarial.
+        :param batch_size: Number of attacks to run simultaneously.
+        :param targeted: True if we should perform a targetted attack, False
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: The typo is still here

File: cleverhans/attacks_tf.py [hunk 34]
```diff
@@ -437,3 +437,245 @@ def jacobian_augmentation(sess, x, X_sub_prev, Y_sub, grads, lmbda,
...
+                         otherwise.
+        :param learning_rate: The learning rate for the attack algorithm.
```
Comment (npapernot) [inline cleverhans/attacks_tf.py]: Was the docstring updated to match the Attacks class?

File: tests_tf/test_attacks.py [hunk 35]
```diff
@@ -275,6 +276,157 @@ def fn(*x, **y):
         self.assertTrue(ok[0])
 
 
+class TestCarliniWagnerL2(CleverHansTest):
+    def setUp(self):
```
Comment (npapernot) [inline tests_tf/test_attacks.py]: The np seed seems to be missing

**Comment by npapernot** [comment]
I left a few comments, but looks like it's ready to merge afterwards

**Comment by npapernot** [comment]
There are also some pep8 errors left:

```
/home/travis/build/tensorflow/cleverhans/cleverhans/attacks.py:604:1: W293 blank line contains whitespace
/home/travis/build/tensorflow/cleverhans/cleverhans/attacks.py:614:73: W291 trailing whitespace
/home/travis/build/tensorflow/cleverhans/cleverhans/attacks.py:615:77: W291 trailing whitespace
```

File: cleverhans/attacks.py (hunk 31 above)
Comment (carlini) [inline cleverhans/attacks.py]: I commented above about this.

File: tests_tf/test_attacks.py (hunk 35 above)
Comment (carlini) [inline tests_tf/test_attacks.py]: It's done in the super class.
